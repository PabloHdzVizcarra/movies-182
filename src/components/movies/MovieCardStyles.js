import styled from "styled-components";
import { respondTo } from "../../styles/_respondTo";

export const ContainCard = styled.div`
  display: grid;
  flex-direction: column;
  background-color: rgba(82, 72, 156, .1);
  border-radius: 6px;
  font-family: 'Roboto', sans-serif;
  width: 100%;
  padding-bottom: 10px;

  h4 {
    font-family: 'Roboto', sans-serif;
    font-weight: 900;
    font-size: 20px;
    color: #ff6d00;
    margin: 5px;
  }

  img {
    border-radius: 6px 6px 0 0;
    width: 100%;
    -webkit-box-shadow: 10px 9px 10px -5px rgba(0, 0, 0, 0.75);
    -moz-box-shadow: 10px 9px 10px -5px rgba(0, 0, 0, 0.75);
    box-shadow: 10px 9px 10px -5px rgba(0, 0, 0, 0.75);
  }

  svg {
    font-size: 25px;
  }

  ${respondTo.sm`
    img {
      height:550px;
    }
  `}
  ${respondTo.md`
    img {
      height: 360px;
    }
  `}
  ${respondTo.lg`
    img {
      height: 315px;
    }
  `}

  ${respondTo.xg`
    img {
      height: 380px;
    }
  `}

  
`;

export const IconAndText = styled.div`
  display: flex;
  align-items: center;
  margin-top: 2px;

  p {
    margin: 2px;
    font-size: 14px;
    font-weight: 400;
  }

  div {
    display: flex;
    margin-left: 5px;
  }
`;

export const TitleAndIcon = styled.div`
  display: flex;
  align-items: baseline;
  justify-content: space-between;

  svg {
    &:hover {
      color: #ff1a1a;
      cursor: pointer;
    }
  }
`;

export const InfoData = styled.div`
  padding-left: 5px;
  padding-right: 5px;
  width: 100%;

`;

export const TextDeleteMovie = styled.p`
  margin: 0;
  margin-left: 7px;
  font-size: 15px;
  color: #ff3333;
  font-weight: bold;
  cursor: pointer;
  width: fit-content;
  margin-top: -5px;

  transition: ease color 300ms;

    &:hover {
      color: #D50000;
    }
`;