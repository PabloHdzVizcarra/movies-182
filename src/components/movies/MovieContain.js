import React from 'react'

export const MovieContain = () => {
  return (
    <div>
      
    </div>
  )
}
