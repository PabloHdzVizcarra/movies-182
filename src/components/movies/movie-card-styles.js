import styled from "styled-components";

export const ContainCard = styled.div`
  display: flex;
  flex-direction: column;
  padding: 1rem;
  background-color: rgba(82, 72, 156, .1);
  border-radius: 6px;
  /* height: fit-content; */

  h4 {
    margin-top: 5px;
    margin-bottom: 5px;
    font-family: "Comfortaa", cursive;
    font-size: 20px;
  }

  img {
    width: 100%;
    -webkit-box-shadow: 10px 9px 10px -5px rgba(0, 0, 0, 0.75);
    -moz-box-shadow: 10px 9px 10px -5px rgba(0, 0, 0, 0.75);
    box-shadow: 10px 9px 10px -5px rgba(0, 0, 0, 0.75);
  }
`;

export const IconAndText = styled.div`
  display: flex;
  align-items: center;
  margin-top: 5px;

  p {
    margin: 2px;
    font-weight: bold;
    font-family: monospace;
  }

  div {
    display: flex;
    margin-left: 10px;
    font-size: 13px;
  }
`;

export const TitleAndIcon = styled.div`
  display: flex;
  align-items: baseline;
  justify-content: space-between;

  svg {
    &:hover {
      color: #ff1a1a;
      cursor: pointer;
    }
  }
`;

export const InfoData = styled.div`
  
`;