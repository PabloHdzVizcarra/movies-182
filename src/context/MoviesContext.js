import React, { createContext, useReducer, useCallback, useState } from 'react';
import { moviesReducer } from '../reducers/moviesReducer';
import { types } from '../types/types';
import { db, firebase } from '../libs/firebase';

export const MoviesContext = createContext();



export const MoviesProvider = ({ children }) => {

  const [state, dispatch] = useReducer(moviesReducer, {
    movies: [],
    favoriteMovies: [],
    searchMovie: [],
    movie: {},
    loadingMovies: true
  });

  const [errorForm, setErrorForm] = useState({
    error: false,
    message: ''
  });
  
  const { movies, favoriteMovies, searchMovie, movie, loadingMovies } = state;
  

  const getTopRatedMovies = useCallback(async () => {
    try {
      const apiKey = process.env.REACT_APP_APIKEY;
      
      console.log('OBTENIENDO PELICULAS');
      const data = await fetch(`https://api.themoviedb.org/3/movie/top_rated?api_key=${apiKey}&language=es-MX&page=1`);
      const { results } = await data.json();
      
      dispatch({
        type: types.loadingTopRatedMovies,
        payload: results
      });
    } catch (error) {
      console.log(error);
    }
  }, []);

  const getPopularMovies = useCallback(async () => {
    try {
      const apiKey = process.env.REACT_APP_APIKEY;
      
      console.log('OBTENIENDO PELICULAS');
      const data = await fetch(`https://api.themoviedb.org/3/movie/popular?api_key=${apiKey}&language=es-MX&page=1`);
      const { results } = await data.json();
      
      dispatch({
        type: types.loadingPopularMovies,
        payload: results
      });
    } catch (error) {
      console.log(error);
    }
  }, []);

  const getUpcomingMovies = useCallback(async () => {

    try {
      const apiKey = process.env.REACT_APP_APIKEY;
      
      console.log('OBTENIENDO PELICULAS');
      const data = await fetch(`https://api.themoviedb.org/3/movie/upcoming?api_key=${apiKey}&language=es-MX&page=1`);
      const { results } = await data.json();
      
      dispatch({
        type: types.loadingUpcomingMovies,
        payload: results
      });
    } catch (error) {
      console.log(error);
    }
  }, []);

  const getMoviesFirebase = async(uid) => {

    try {
      
      let allFavoritesMovies = [];

      await db.collection(uid)
      .doc("movies")
      .collection("favorites")
      .get()
      .then((querySnapshot) => {
        querySnapshot.forEach((doc) => {
          let movie;
          movie = { ...doc.data() }
          movie.docID = doc.id;
          allFavoritesMovies.push(movie)
        });
      });

      // const reversedArray = allFavoritesMovies.reverse();

      dispatch({
        type: types.getFavoritesMoviesFirebase,
        payload: allFavoritesMovies
      })

    } catch (error) {
      console.log(error);
    }
  }

  const getMovieSearch = async (key) => {

    try {
      const apiKey = process.env.REACT_APP_APIKEY;
      const data = await fetch(`https://api.themoviedb.org/3/search/movie?api_key=${apiKey}&query=${key}`);
      
      const { results } = await data.json();
      console.log(results);

      if (results.length === 0) {
        return setErrorForm({
          error: true,
          message: 'Intenta buscando otra pelicula'
        })
      }

      setErrorForm({
        error: false,
        message: ''
      })

      dispatch({
        type: types.searchMovies,
        payload: results
      });

    } catch (error) {
      console.log(error);
    }
  }

  const getMovieID = async (movieID) => {

    try {

      const apiKey = process.env.REACT_APP_APIKEY;
      const data = await fetch(`https://api.themoviedb.org/3/movie/${movieID}?api_key=${apiKey}&language=es-ES`);

      const resp = await data.json();
      
      dispatch({
        type: types.selectedMovie,
        payload: resp
      })
      
    } catch (error) {
      console.log(error);
    }
  }



  const addFavoriteMovie = async (movie) => {

    const {
      poster_path,
      id,
      genres,
      original_title,
      original_language,
      overview,
      popularity,
      production_countries,
      release_date,
      tagline,
      title,
      vote_average,
      vote_count
    } = movie;

    const imgMovie = `https://image.tmdb.org/t/p/w500/${poster_path}`;

    const favoriteMovie = {
      imgMovie,
      id,
      genres,
      original_title,
      original_language,
      overview,
      popularity,
      production_countries,
      release_date,
      tagline,
      title,
      vote_average,
      vote_count
    }

    const { uid } = firebase.auth().currentUser;

    if (uid) {

      const { id } = await db.collection(uid).doc('movies').collection('favorites').add(favoriteMovie);

      favoriteMovie.firebaseID = id;

      dispatch({
        type: types.addFavoritesMovies,
        payload: favoriteMovie
      })
      
      console.log('PELICULA AGREGADA A FAVORITOS');

    } else {
      throw new Error('No hay usuario autenticado')
    }

  }

  const deleteMovieFromFavorites = async (userID, docID) => {
    console.log(docID);

    dispatch({
      type: types.removeMovieFromFavorites,
      payload: docID
    })
    

    await db.collection(userID).doc('movies').collection('favorites').doc(docID)
      .delete();

    return console.log(`PELICULA ELIMINADA`);
  }
  
  return (
    <MoviesContext.Provider value={{
      movie,
      movies,
      favoriteMovies,
      searchMovie,
      loadingMovies,
      errorForm,
      getMovieID,
      getPopularMovies,
      getTopRatedMovies,
      getUpcomingMovies,
      getMovieSearch,
      addFavoriteMovie,
      getMoviesFirebase,
      deleteMovieFromFavorites,
      dispatch
    }}>
      {children}
    </MoviesContext.Provider>
  )
}