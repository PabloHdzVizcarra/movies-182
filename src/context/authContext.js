import React from 'react'
import { types } from '../types/types';

const AuthStateContext = React.createContext();
const AuthDispatchContext = React.createContext();

function authReducer(state, action) {
  switch (action.type) {

    case types.setActiveUser:
      return {
        ...state,
        activeUser: action.payload,
        isActived: true
      }

    case types.logoutUser:
      return {
        ...state,
        activeUser: {},
        isActived: false
      }
    default: {
      throw new Error(`Unhandled action type: ${action.type}`)
    }
  }
}


function AuthProvider({ children }) {
  const [state, dispatch] = React.useReducer(authReducer, {
    activeUser: {},
    isActived: false
  });

  return (
    <AuthStateContext.Provider value={state}>
      <AuthDispatchContext.Provider value={dispatch}>
        {children}
      </AuthDispatchContext.Provider>
    </AuthStateContext.Provider>
  )
}

function useAuthState() {
  const context = React.useContext(AuthStateContext);

  if (context === undefined) {
    throw new Error('useAuthState must be used within AuthProvider')
  }

  return context
}

function useAuthDispatch() {
  const context = React.useContext(AuthDispatchContext);

  if (context === undefined) {
    throw new Error('useAuthDispatch must be used within AuthProvider')
  }

  return context
}

export {
  AuthProvider,
  useAuthState,
  useAuthDispatch
}