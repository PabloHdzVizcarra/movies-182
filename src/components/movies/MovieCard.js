import React, { useContext } from "react";
import styled from "styled-components";
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { Link } from "react-router-dom";
import { useAuthState } from "../../context/authContext";
import { MoviesContext } from "../../context/MoviesContext";



const ContainCard = styled.div`
  display: flex;
  flex-direction: column;
  padding: 1rem;
  background-color: rgb(0, 0, 0, .1);
  border-radius: 6px;

  h4 {
    margin-top: 5px;
    margin-bottom: 5px;
    font-family: 'Comfortaa', cursive;
    font-size: 20px;
  }

  img {
    width: 100%;
    -webkit-box-shadow: 10px 9px 10px -5px rgba(0,0,0,0.75);
    -moz-box-shadow: 10px 9px 10px -5px rgba(0,0,0,0.75);
    box-shadow: 10px 9px 10px -5px rgba(0,0,0,0.75);
  }

`;

const IconAndText = styled.div`
  display: flex;
  align-items: center;
  margin-top: 5px;
  
  p {
    margin: 2px;
    font-weight: bold;
    font-family: monospace;
  }
`;

const TitleAndIcon = styled.div`
  display: flex;
  align-items: baseline;
  justify-content: space-between;

  svg {
    &:hover {
      cursor: pointer;
    }
  }
`;

export const MovieCard = ({ docID, isActived, id, title, poster_path, imgMovie, popularity, release_date, vote_average }) => {

  const { activeUser } = useAuthState();
  const { deleteMovieFromFavorites } = useContext(MoviesContext);

  function handleDeleteFavoriteMovie() {
    deleteMovieFromFavorites(activeUser.uid, docID);


  }

  return (
    <ContainCard>
      <TitleAndIcon>
        <h4>{title}</h4>
        {isActived
          && <FontAwesomeIcon
          icon="heart-broken" 
          onClick={handleDeleteFavoriteMovie}
          />
        }
      </TitleAndIcon>
      <Link to={`/search/${id}`} >
        {
          imgMovie
            ? <img src={imgMovie} alt={title}/>
            : <img src={`https://image.tmdb.org/t/p/w342/${poster_path}` || ''} alt="No hay Imagen Disponible" />
        }
      </Link>
      <IconAndText>
        <FontAwesomeIcon icon="fire" color="#ff9900"/>
        <p>{popularity}</p>
      </IconAndText>
      <IconAndText>
        <FontAwesomeIcon icon="thumbs-up" color="#b98946" />
        <p>{vote_average}</p>
      </IconAndText>
      <IconAndText>
        <FontAwesomeIcon icon="calendar-day" color="#1a75ff"/>
        <p>{release_date}</p>
      </IconAndText>
    </ContainCard>
  )
};
