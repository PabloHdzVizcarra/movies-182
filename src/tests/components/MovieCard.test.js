import React from "react";
import "@testing-library/jest-dom";
import { mount } from "enzyme";
import { MovieCard } from "../../components/movies/MovieCard";
import { AuthProvider } from "../../context/authContext";
import { MoviesContext } from "../../context/MoviesContext";
import { MemoryRouter } from "react-router-dom";

describe("Test in <MovieCard />", () => {
  const contextValue = {
    deleteMovieFromFavorites: jest.fn(),
  };

  const wrapper = mount(
    <AuthProvider>
      <MoviesContext.Provider value={contextValue}>
        <MemoryRouter>
          <MovieCard
            title="Spiderman"
            docID="123456"
            id="123"
            poster_path="imagen"
            release_date="2020-01-20"
            vote_average="7.5"
            popularity={"2010"}
            isActived={true}
          />
        </MemoryRouter>
      </MoviesContext.Provider>
    </AuthProvider>
  );

  test("should show correctly component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should if the user is logged in, it must show the paragraph with the text "Eliminar pelicula" ', () => {
    const textDelete = wrapper.find(".sc-AxhUy").text().trim();
    expect(textDelete).toBe("Eliminar pelicula");
  });

  test('when the user is not logged in, the text "Eliminar pelicula should not be displayed', () => {
    const wrapper = mount(
      <AuthProvider>
        <MoviesContext.Provider value={contextValue}>
          <MemoryRouter>
            <MovieCard
              title="Spiderman"
              docID="123456"
              id="123"
              poster_path="imagen"
              release_date="2020-01-20"
              vote_average="7.5"
              popularity={"2010"}
              isActived={false}
            />
          </MemoryRouter>
        </MoviesContext.Provider>
      </AuthProvider>
    );

    const textDelete = wrapper.find(".sc-AxhUy").exists();
    expect(textDelete).toBe(false);
  });

  test('click on the "Eliminar pelicula" paragraph to call the deleteMovieFromFavorites function ', () => {
    wrapper.find(".sc-AxhUy").simulate("click");
    expect(contextValue.deleteMovieFromFavorites).toBeCalled();
  });
});
