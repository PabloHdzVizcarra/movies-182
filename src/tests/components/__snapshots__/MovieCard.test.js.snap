// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <MovieCard /> should show correctly component 1`] = `
<div
  className="sc-AxjAm kuIkN animate__animated animate__fadeIn"
>
  <div>
    <a
      href="/search/123"
      onClick={[Function]}
    >
      <img
        alt="No hay Imagen Disponible"
        src="https://image.tmdb.org/t/p/w342/imagen"
      />
    </a>
  </div>
  <div>
    <div
      className="sc-AxhCb buhHwi"
    >
      <h4>
        Spiderman
      </h4>
      <div
        className="sc-AxirZ dZnDSZ"
      >
        <div>
          <p>
            Calificacion:
          </p>
          <p>
            <b>
              7.5
            </b>
             
          </p>
        </div>
      </div>
      <div
        className="sc-AxirZ dZnDSZ"
      >
        <div>
          <p>
             Likes: 
          </p>
          <p>
            <b>
              2010
            </b>
             
          </p>
        </div>
      </div>
      <div
        className="sc-AxirZ dZnDSZ"
      >
        <div
          style={
            Object {
              "paddingBottom": "10px",
            }
          }
        >
          <p>
            Estreno: 
          </p>
          <p>
            <b>
              20 enero 2020
            </b>
             
          </p>
        </div>
      </div>
      <p
        className="sc-AxhUy KsMZo"
        onClick={[Function]}
      >
        Eliminar pelicula
      </p>
    </div>
  </div>
</div>
`;
